File: Phan_4/chon_hoat_dong.py
def tuong_thich(h1, h2):
    return h1[2] <= h2[1] or h2[2] <= h1[1]

def tieu_chi_ket_thuc_som(hd):
    ds = sorted(hd, key=lambda x: (x[2], x[1]))
    chon = []
    cur_f = 0
    for h in ds:
        if h[1] >= cur_f:
            chon.append(h[0])
            cur_f = h[2]
    return chon

def tieu_chi_bat_dau_som(hd):
    ds = sorted(hd, key=lambda x: (x[1], x[2]))
    chon = []
    for h in ds:
        if all(tuong_thich(h, next(x for x in hd if x[0] == cid)) for cid in chon):
            chon.append(h[0])
    return chon

def tieu_chi_ngan_nhat(hd):
    ds = sorted(hd, key=lambda x: (x[2] - x[1], x[2]))
    chon = []
    for h in ds:
        if all(tuong_thich(h, next(x for x in hd if x[0] == cid)) for cid in chon):
            chon.append(h[0])
    return chon

def tieu_chi_it_chong_lan_nhat(hd):
    def dem_xung_dot(target, tap):
        return sum(1 for h in tap if h[0] != target[0] and not tuong_thich(target, h))
    
    con_lai = list(hd)
    chon = []
    while con_lai:
        best = min(con_lai, key=lambda x: (dem_xung_dot(x, con_lai), x[2]))
        chon.append(best[0])
        con_lai = [h for h in con_lai if h[0] != best[0] and tuong_thich(best, h)]
    return chon

def kiem_chung_toi_uu(hd):
    import itertools
    for k in range(len(hd), 0, -1):
        for comb in itertools.combinations(hd, k):
            ok = True
            for i in range(len(comb)):
                for j in range(i + 1, len(comb)):
                    if not tuong_thich(comb[i], comb[j]):
                        ok = False
                        break
                if not ok:
                    break
            if ok:
                return [h[0] for h in comb]
    return []

if __name__ == "__main__":
    hd = [
        ("H1", 1, 5), ("H2", 2, 5), ("H3", 2, 6), ("H4", 3, 4), ("H5", 4, 8),
        ("H6", 6, 9), ("H7", 8, 11), ("H8", 9, 14), ("H9", 11, 13), ("H10", 12, 15)
    ]

    c1 = tieu_chi_ket_thuc_som(hd)
    c2 = tieu_chi_bat_dau_som(hd)
    c3 = tieu_chi_ngan_nhat(hd)
    c4 = tieu_chi_it_chong_lan_nhat(hd)
    opt = kiem_chung_toi_uu(hd)

    print("=== KET QUA CHON HOAT DONG (PYTHON) ===")
    print(f"Ket thuc som nhat:       {{{', '.join(c1)}}} (So luong: {len(c1)}) -> {'CO' if len(c1)==len(opt) else 'KHONG'}")
    print(f"Bat dau som nhat:        {{{', '.join(c2)}}} (So luong: {len(c2)}) -> {'CO' if len(c2)==len(opt) else 'KHONG'}")
    print(f"Ngan nhat:               {{{', '.join(c3)}}} (So luong: {len(c3)}) -> {'CO' if len(c3)==len(opt) else 'KHONG'}")
    print(f"It chong lan nhat:       {{{', '.join(c4)}}} (So luong: {len(c4)}) -> {'CO' if len(c4)==len(opt) else 'KHONG'}")
    print(f"So nhieu nhat (Toi uu):  {{{', '.join(opt)}}} (So luong: {len(opt)}) -> CO")

File: Phan_4/chon_hoat_dong.jl
# Quy uoc: Trong Julia mang danh chi so bat dau tu 1

struct HoatDong
    id::String
    s::Int
    f::Int
end

tuong_thich(a::HoatDong, b::HoatDong) = (a.f <= b.s || b.f <= a.s)

function ket_thuc_som(ds)
    # Sap xep theo f tang dan
    ds_sorted = sort(ds, by = x -> (x.f, x.s))
    chon = String[]
    cur_f = 0
    for h in ds_sorted
        if h.s >= cur_f
            push!(chon, h.id)
            cur_f = h.f
        end
    end
    return chon
end

function bat_dau_som(ds)
    # Sap xep theo s tang dan
    ds_sorted = sort(ds, by = x -> (x.s, x.f))
    chon = HoatDong[]
    for h in ds_sorted
        if all(tuong_thich(h, c) for c in chon)
            push!(chon, h)
        end
    end
    return [h.id for h in chon]
end

function ngan_nhat(ds)
    # Sap xep theo thoi luong (f - s) tang dan
    ds_sorted = sort(ds, by = x -> (x.f - x.s, x.f))
    chon = HoatDong[]
    for h in ds_sorted
        if all(tuong_thich(h, c) for c in chon)
            push!(chon, h)
        end
    end
    return [h.id for h in chon]
end

function it_chong_lan(ds)
    # Chon hoat dong it xung dot nhat voi cac hoat dong con lai
    con_lai = copy(ds)
    chon = String[]
    while !isempty(con_lai)
        dem_xung_dot(target) = count(h -> h.id != target.id && !tuong_thich(target, h), con_lai)
        sort!(con_lai, by = x -> (dem_xung_dot(x), x.f))
        best = con_lai[1]
        push!(chon, best.id)
        filter!(h -> h.id != best.id && tuong_thich(best, h), con_lai)
    end
    return chon
end

# Kiem chung doc lap: Vet can moi tap con tu lon xuong nho
function kiem_chung_toi_uu(ds)
    n = length(ds)
    best = String[]
    for mask in 1:(1 << n) - 1
        sub = [ds[i] for i in 1:n if (mask & (1 << (i - 1))) != 0]
        ok = true
        for i in 1:length(sub)
            for j in (i + 1):length(sub)
                if !tuong_thich(sub[i], sub[j])
                    ok = false
                    break
                end
            end
            !ok && break
        end
        if ok && length(sub) > length(best)
            best = [h.id for h in sub]
        end
    end
    return best
end

ds = [
    HoatDong("H1", 1, 5), HoatDong("H2", 2, 5), HoatDong("H3", 2, 6), HoatDong("H4", 3, 4), HoatDong("H5", 4, 8),
    HoatDong("H6", 6, 9), HoatDong("H7", 8, 11), HoatDong("H8", 9, 14), HoatDong("H9", 11, 13), HoatDong("H10", 12, 15)
]

c1 = ket_thuc_som(ds)
c2 = bat_dau_som(ds)
c3 = ngan_nhat(ds)
c4 = it_chong_lan(ds)
opt = kiem_chung_toi_uu(ds)

println("=== KET QUA CHON HOAT DONG (JULIA) ===")
println("Ket thuc som nhat:       {", join(c1, ", "), "} (So luong: $(length(c1))) -> $(length(c1) == length(opt) ? "CO" : "KHONG")")
println("Bat dau som nhat:        {", join(c2, ", "), "} (So luong: $(length(c2))) -> $(length(c2) == length(opt) ? "CO" : "KHONG")")
println("Ngan nhat:               {", join(c3, ", "), "} (So luong: $(length(c3))) -> $(length(c3) == length(opt) ? "CO" : "KHONG")")
println("It chong lan nhat:       {", join(c4, ", "), "} (So luong: $(length(c4))) -> $(length(c4) == length(opt) ? "CO" : "KHONG")")
println("So nhieu nhat (Toi uu):  {", join(opt, ", "), "} (So luong: $(length(opt))) -> CO")